### PR TITLE
feat: Compute per-query average for 2D `retrieval_normalized_dcg`

### LINKED ISSUE #3216: `retrieval_normalized_dcg` should compute per-query average when given 2D inputs (IR-standard behavior)
## 🐛 Bug
Currently, when passing a 2D tensor (`[num_queries, num_documents]`) to `retrieval_normalized_dcg`, the function flattens both `preds` and `target` and computes DCG/IDCG on the concatenated list.
This treats all queries as a single large ranking problem.

In Information Retrieval (IR) and recommender systems, the standard practice for NDCG is:  
- Compute NDCG per query
- Then take the macro average over queries  

Flattening across queries changes the interpretation of the metric and can lead to inflated or misleading results.

### To Reproduce

```python
from torchmetrics.functional.retrieval import retrieval_normalized_dcg
import torch

# Query 1
p1 = retrieval_normalized_dcg(torch.tensor([0.1, 0.2, 0.3]), torch.tensor([0, 1, 0]))
print(p1)  # tensor(0.6309)

# Query 2
p2 = retrieval_normalized_dcg(torch.tensor([0.8, 0.1, 0.05]), torch.tensor([1, 0, 0]))
print(p2)  # tensor(1.0000)

print("Mean per-query NDCG:", (p1 + p2) / 2)
# tensor(0.8155)

# Batched input (2D)
p_batch = retrieval_normalized_dcg(
    torch.tensor([[0.1, 0.2, 0.3], [0.8, 0.1, 0.05]]),
    torch.tensor([[0, 1, 0], [1, 0, 0]]),
)
print("Batch NDCG:", p_batch)
# tensor(0.9197) <-- Not the mean per-query value
```

Here, the batch value 0.9197 is different from the expected per-query average 0.8155 because the function flattens both queries before computing NDCG.

**Environment**
- macOS 15.4.1 (Sequoia) on Intel MacBook Pro
- Python 3.12.6
- torch==2.2.0
- torchmetrics==1.8.1

### PR DESCRIPTION
**Pull request by rintaro121**
## What does this PR do?

Fixes #3216 

<details>
  <summary>Before submitting</summary>

- [x] Was this **discussed/agreed** via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/torchmetrics/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [ ] Did you make sure to **update the docs**?
- [x] Did you write any new **necessary tests**?

</details>

<details>
  <summary>PR review</summary>

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

</details>

## Motivation
Currently, when passing a 2D tensor ([num_queries, num_documents]) to `retrieval_normalized_dcg`, the function flattens both preds and target and computes DCG/IDCG on the concatenated list.
This treats all queries as a single large ranking problem.

```py
# To reproduce
from torchmetrics.functional.retrieval import retrieval_normalized_dcg
import torch

# Query 1
p1 = retrieval_normalized_dcg(torch.tensor([0.1, 0.2, 0.3]), torch.tensor([0, 1, 0]))
print(p1)  # tensor(0.6309)

# Query 2
p2 = retrieval_normalized_dcg(torch.tensor([0.8, 0.1, 0.05]), torch.tensor([1, 0, 0]))
print(p2)  # tensor(1.0000)

print("Mean per-query NDCG:", (p1 + p2) / 2)
# tensor(0.8155)

# Batched input (2D)
p_batch = retrieval_normalized_dcg(
    torch.tensor([[0.1, 0.2, 0.3], [0.8, 0.1, 0.05]]),
    torch.tensor([[0, 1, 0], [1, 0, 0]]),
)
print("Batch NDCG:", p_batch)
# tensor(0.9197) <-- Not the mean per-query value
```
## Changes
Fixes to `retrieval_normalized_dcg`:
- Updated the logic to correctly handle 2D tensors as batches of samples.
- Added a new argument empty_target_action to specify behavior when a target contains no positives:(This argument has been added because the presence of positive labels in target is not guaranteed.)
  - "skip" (default): Exclude the sample from the average computation.
  - "pos": Treat the score as 1.0.
  - "neg": Treat the score as 0.0.

Test additions and modifications:
- Added new test cases covering the above fixes and feature additions.
- Added tests to verify that each option of empty_target_action ("skip", "pos", "neg") works correctly.
- Defined input tensors for use in the new tests.
- Adjusted the arguments (e.g., k=...) passed to scikit-learn's `ndcg_score` to align conditions with our implementation.

## For the reviewer
Feedback on the naming and default behavior of the new `empty_target_action` argument would be greatly appreciated. Please also let us know if any test coverage is missing.

## Did you have fun?
Make sure you had fun coding 🙃

----
📚 Documentation preview 📚: https://torchmetrics--3229.org.readthedocs.build/en/3229/

### CODE REVIEW

File: src/torchmetrics/functional/retrieval/ndcg.py [hunk 1]
```diff
@@ -95,19 +131,37 @@ def retrieval_normalized_dcg(preds: Tensor, target: Tensor, top_k: Optional[int]
...
+    original_shape = preds.shape
     preds, target = _check_retrieval_functional_inputs(preds, target, allow_non_binary_target=True)
 
-    top_k = preds.shape[-1] if top_k is None else top_k
+    # reshape back if input was 2D
+    if len(original_shape) == 2:
+        preds = preds.view(original_shape)
+        target = target.view(original_shape)
+    else:
+        preds = preds.unsqueeze(0)
```
[the remark this answers is not in the document]
Comment (rintaro121) [reply src/torchmetrics/functional/retrieval/ndcg.py:142]: This branch handles 1D inputs and `target.unsqueeze(0)` is correct to match the `preds.unsqueeze(0)` on the line above it. The tests correctly pass with this implementation.

**Comment by rintaro121** [comment]
@justusschock 
Friendly ping on this PR.

Just to give a quick update on the changes:

- I've added support for 2D tensor inputs (batched queries) in retrieval_normalized_dcg.
- Introduced the empty_target_action parameter to handle queries with no relevant docs, matching other retrieval metrics.
- I've also recently updated the code to address the review comments from Copilot.

Also, I noticed that the CI checks seem to have timed out. Could a maintainer help re-trigger them when you have a moment, or let me know if there's anything I need to fix on my end?

Happy to make any further adjustments based on your feedback. Thanks🙏
